### PR TITLE
Uni v3 twap

### PR DESCRIPTION
**Pull request by brianmcmichael**

### CODE REVIEW

File: src/KilnUniV3.sol
Comment (gbalabasquer) [inline src/KilnUniV3.sol:69]: Why not make this manageable or at least as constructor param?

File: src/KilnUniV3.sol
Comment (gbalabasquer) [inline src/KilnUniV3.sol]: This is only true if `yen == WAD`

File: src/KilnUniV3.sol
Comment (gbalabasquer) [inline src/KilnUniV3.sol]: `part` or `multiplier`? as actually depending on the value of `yen` we could ask for greater, equal or less than the avg price.

File: src/KilnMom.sol
Comment (gbalabasquer) [inline src/KilnMom.sol]: Hmmm I think the `dst` should be preset by governance with delay. Then it is not possible to "steal" anything without passing the governance delay.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:69]: It is fileable.
The default values for `scope` and `yen` are not mandatory, but are a safety measure (as leaving them 0 by mistake would work, but with a more risky configuration).

File: src/KilnMom.sol
Comment (rockyfour) [reply src/KilnMom.sol]: What do you think about making it immutable (for sake of making the mom as simple as possible)?
In the unlikely case that needs to change then a new mom should be deployed and authed on kiln.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol]: I think it means to continue the previous sentence about being enabled by default.
Can be made clearer.

File: src/KilnUniV3.sol
Comment (gbalabasquer) [reply src/KilnUniV3.sol:69]: Yes sorry, just got confused with the comment in the constructor and didn't check it well.
"Fire will perform the trade only when the amount of tokens received is better than the 1 hour average price."
Maybe I'd changing it to something like: "With the initial values, fire will perform the trade only when the amount of tokens received is equal or better than the 1 hour average price."

File: src/KilnUniV3.sol
Comment (gbalabasquer) [reply src/KilnUniV3.sol]: Yep I just commented how we could improve this message to avoid any confusion.

File: src/KilnMom.sol
Comment (gbalabasquer) [reply src/KilnMom.sol]: I'm fine making it immutable as well.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol]: changed on [e8c0964](https://github.com/makerdao/dss-kiln/pull/9/commits/e8c0964fe1a80e828163e3aad9a8caaf22e3fc9e)

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol]: changed on [e8c0964](https://github.com/makerdao/dss-kiln/pull/9/commits/e8c0964fe1a80e828163e3aad9a8caaf22e3fc9e)

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:69]: changed on [e8c0964](https://github.com/makerdao/dss-kiln/pull/9/commits/e8c0964fe1a80e828163e3aad9a8caaf22e3fc9e)

File: src/KilnMom.sol
Comment (rockyfour) [reply src/KilnMom.sol]: Done on [74d16c3](https://github.com/makerdao/dss-kiln/pull/9/commits/74d16c3ad6ed2a52642dca35b43360cb54cbe5c6)

File: src/KilnUniV3.sol
Comment (gbalabasquer) [reply src/KilnUniV3.sol]: Looks good!

File: src/KilnMom.sol
Comment (gbalabasquer) [reply src/KilnMom.sol]: Looks good!

File: .editorconfig
Comment (iamchrissmith) [inline .editorconfig]: duplicated

File: .gitignore
Comment (iamchrissmith) [inline .gitignore]: pretty sure we prefer to have IDE specific ignores in on the local machine's global ignore rather than in each repo's gitignore.  If so, we should remove this and the `.vscode` one

File: README.md
Comment (iamchrissmith) [inline README.md]: love the meme!

Comment (iamchrissmith) [inline README.md]: Instead of 1000000000000000000 would be better to write as value the percentage * WAD like you do below in the examples

Comment (iamchrissmith) [reply README.md]: Might also be good to explain the difference between leaving the default (0% slippage/discount) and setting it to 0.  As I understand it: 

`file('yen', WAD) // default`: Trade when you can get the same price as the last hour TWAP average (accounting for slippage)

`file('yen', 0)`: Just dump at whatever price

File: src/KilnBase.sol
Comment (iamchrissmith) [inline src/KilnBase.sol:3]: Are we moving away from having the title/desc in the comments here?

File: src/KilnBase.sol
Comment (iamchrissmith) [inline src/KilnBase.sol]: should this be `lock` too? I know its auth, but could there be instances where `fire` triggers an `auth` contract to try to `rug`?

File: src/KilnBase.sol
Comment (iamchrissmith) [inline src/KilnBase.sol:34]: I *think* we want `immutable`s above other variables iirc

File: src/KilnBase.sol
Comment (iamchrissmith) [inline src/KilnBase.sol:108]: Kurt suggested a change in d3m that I think applies here too:

```suggestion
        require(block.timestamp >= zzz, "KilnBase/fired-too-soon");
        uint256 amt = _min(GemLike(sell).balanceOf(address(this)), lot);
        require(amt > 0, "KilnBase/no-balance");
        uint256 swapped = _swap(amt);
        zzz = block.timestamp + hop;
```

File: src/KilnMom.sol
Comment (iamchrissmith) [inline src/KilnMom.sol:81]: I kind of think `_dst` should be a parameter of the `rug` function.  There are some scenarios where we might set a `dst` when creating the mom and then when we need to call it our `dst` is not safe and we would want to send things someplace else. Or where we want different `dst` for different `who`

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol:52]: This line probably makes more sense above the `file` function for `yen`

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol]:
```suggestion
    // @param _sell                the contract address of the token that will be sold
    // @param _buy                the contract address of the token that will be purchased
    // @param _uniV3Router  the address of the current Uniswap V3 swap router
    // @param _receiver         the address of the account which will receive the funds to be bought
```

(I think that aligns the spacing... always hard to tell in the GH comments)

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol]:
```suggestion
interface SwapRouterLike {
     // https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/SwapRouter.sol#L132
    function exactInput(ExactInputParams calldata params) external returns (uint256 amountOut);
```

Since this points to a specific function in their code and the `factory()` piece is in an inherited contract

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol:73]: I think this should go above the constructor

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol:97]:
```suggestion
            // filing lot or hop from KilnBase
            super.file(what, data);
```

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol:109]: I find this a little easier to consume:
```suggestion
        uint256 amountMin = (_yen == 0) ? 0 : quote(_path, amount, uint32(scope)) * _yen / WAD;
```

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol]:
```suggestion
        bytes    memory _path = path;
        uint256               _yen  = yen;
```

File: src/KilnUniV3.sol
Comment (iamchrissmith) [inline src/KilnUniV3.sol:104]: There is not a way that Uniswap doesn't complete all of this trade without reverting correct? (want to make sure we don't have any dangling approvals)

File: src/KilnUniV3.t.sol
Comment (iamchrissmith) [inline src/KilnUniV3.t.sol:120]: ❤️

**Review by iamchrissmith — CHANGES_REQUESTED** [review]

File: src/KilnMom.sol
Comment (gbalabasquer) [reply src/KilnMom.sol:81]: This is actually a change I requested (to make `dst` immutable and not define it as param).
`dst` will be the pauseProxy or maybe the `blow` in case of DAI (to send it back to the surplus buffer without selling it).
The problem of making `dst` a param in mom's `rug` is you will be allowing a malicious governance to get the funds without delay. I find that much more risky thing than realizing there is a problem in the previously set recipient.

Comment (iamchrissmith) [reply src/KilnMom.sol:81]: hmm, i see your point, but if we are setting it to the `pauseProxy` anyways, then I'm not sure making it immutable really solves for that except in the cause of DAI going to `blow`.  I don't feel strongly if we feel the immutability is better

File: .editorconfig
Comment (rockyfour) [reply .editorconfig]: removed

File: .gitignore
Comment (rockyfour) [reply .gitignore]: removed

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol:3]: If no objections than yes, I don't think it adds much.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:97]: This comment will require  an update if KilnBase gets changed, so I'd rather not add it.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:109]: I think the current implementation highlights the common path better (and is slightly better gas-wise).
Not feeling that strong about is so let me know if you do.

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol]: Although I can't imagine such a case I see no big harm in adding it for extra safety. Done.

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol:34]: mmm I'm not sure about it (it's not like that in Maker's d3m or TeleportRouter for example).

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:73]: Also not sure about this one. Math stuff is below the constructor in d3m at least.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:104]: Right, in the end of the uniswap `exactInput` call there is a check that the min output amount was retrieved.
Finishing the swap without that check undermines the uniswap logic.
We also know that the entire input amount was used since we use `exactInput`.

File: README.md
Comment (rockyfour) [reply README.md]: Tried to make it clearer in [bea719e](https://github.com/makerdao/dss-kiln/pull/9/commits/bea719ec8334316d6dd84c4a09c642242478a4f3). Let me know if you still think it needs changing.

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol:108]: Mmm what's the motivation for storing the next fireable time instead of the last fire time?

With the way it is now I think updating the value is more straightforward (for example if you want a shorter/longer `hop` it will just take effect without needing to calculate anything special).
With the suggested change, in order to change `hop` and not add any time delay you would need to complicate `file()` with something like:
https://github.com/makerdao/curve-lp-oracle/blob/master/src/CurveLPOracle.sol#L150

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:52]: You are right, but then we'll have to explain also about `scope` and such in the `file` function, which is a bigger change.

TBH I already feel the documentation here (both in the constructor and in the readme) will require a lot of notice and scanning all the wording again if we update something (which is not really desirable).
So would prefer not extending it to one more place (the file functions).

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol]: Ended up removing the specific line reference, so its better accommodating also the other function `factory()`, and aligns better with`ExactInputParams` documentation.

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol]: I think this is now handled in f27eb9172.

File: src/KilnBase.sol
Comment (iamchrissmith) [reply src/KilnBase.sol:3]: works for me

File: src/KilnBase.sol
Comment (iamchrissmith) [reply src/KilnBase.sol:34]: ahh ok, i thought it was the other way :+1:

File: src/KilnBase.sol
Comment (iamchrissmith) [reply src/KilnBase.sol:108]: I'm not sure I see why it needs to complicate the `file`.  The reason we changed it in D3M (it was for the write-off timeline) was so governance couldn't mess with things after the thing was caged.  That is not the case here, but the benefit in my mind is predictability for keepers.  They will know that `fire` happens at a regular interval from the last call and any future governance changes to `hop` will only affect the next `fire`:
Example:
`hop` = 1000
`fire` gets called, `zzz = timestamp + 1000`
keepers know they can call `fire` again in `1000`
in the meantime a spell gets cast that changes `hop` to `10000`
`fire` gets called at the previously set `zzz` and now `zzz = timestamp + 10000`

If we don't make this change then the governance spell affects when the next `fire` occurs and could theoretically reduce the fire and cause `fire` to be called sooner or even in the spell which to me feels like something we should avoid if we can easily do it.

File: src/KilnUniV3.sol
Comment (iamchrissmith) [reply src/KilnUniV3.sol:109]: i don't, lets leave

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol:108]: Well the complication of `file` is for the new `hop` to take effect immediately after it is filed, which is at least the behavior in CurveLpOracle (so no aiming for that kind of keeper predictability there, and at least the keepers developed lately check `fire`ability/`execute`ability each block).

Note that being able to change `hop` immediately allows you to set a very long `hop` and then reduce it if you want (without waiting forever to reach `zzz`). So I thing the ability to call `fire` sooner is a feature, not a bug (and at least is regarded that way in the oracles).

File: src/KilnMom.sol
Comment (rockyfour) [reply src/KilnMom.sol:81]: Not sure I understand the last comment that if `dst` is set to `pauseProxy` then making it immutable (instead of a parameter) does not restrict the `chief`/`authority` from changing `dst` without a time delay.

File: src/KilnBase.sol
Comment (iamchrissmith) [reply src/KilnBase.sol:108]: ok, I didn't see the complication because the benefit I saw to my suggestion was that governance couldn't make the new `hop` take effect immediately (if that is the goal,I agree your approach is the way to go).

I think I disagree and would say we want to prevent governance from having the ability to immediately mess with the fire times, but I don't feel strongly enough to block on this.  Lets leave the comments open and see if others weigh in and have any thoughts

**Comment by iamchrissmith** [comment]
this looks good to send to CS

**Comment by The-Arbiter** [comment]
### KilnBase

* [ ] Solidity version is OK; we could probably bump it - do we want a ceiling for it? 
* [ ] Interfaces match [dss-interfaces](https://github.com/makerdao/dss-interfaces/blob/master/src/ERC/GemAbstract.sol) but are in a different order, OK
* [ ] Wards / auth OK
* [ ] Naming conventions for `hop` `lot` and `zzz` look OK - I would ask why not use `ttl` and `last` but I assume that this is to avoid confusion?
      I would also question the need for `lock` to be public but this is nitpicking. If `lock` prevents reentrancy and changes back to 0 after function execution, what is the benefit of having it as public? The only things that can call and check if it is locked would have to be external code trying to run the `locked` function, which is what `lock` is aiming to prevent in the first place. Internal or private seem to be more appropriate here but again this is a minor thing.
* [ ] Immutable buy / sell addresses OK
* [ ] Events OK
* [ ] Constructor OK
* [ ] Auth modifier OK
* [ ] Lock modifier could save a bunch of gas by using 1 and 2 rather than 0 and 1 (0 is uninit value so costs more). [OZ reentrancyGuard works this way](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol) so it's worth discussing. Edit: it saves about 220 gas per call using 1 and 2 per my testing.
* [ ] `_min` function (marked as internal) is named `min` in other implementations ([autoline](https://github.com/makerdao/dss-auto-line/blob/master/src/DssAutoLine.sol#L68) ) but `_min` in others ([teleport](https://github.com/makerdao/dss-teleport/blob/master/src/TeleportJoin.sol)); what one do we want to go with? Internal should be preceded by a `_` but there is some inconsistency.
      Also, the `_min` here doesn't use the `z` return var. We should change the code from `return x <= y ? x : y;` to `z = x <= y ? x : y;` (from teleport) or have an anonymous return variable.
* [ ] `rely` and `deny` code is fine - whether we want the same formatting as older files (multi line) or not is up to you.
* [ ] `file` code is fine but it's been marked as `external` in the past ([teleport](https://github.com/makerdao/dss-teleport/blob/master/src/TeleportRouter.sol#L85) and [IAM](https://github.com/makerdao/dss-auto-line/blob/master/src/DssAutoLine.sol#L68). I am assuming that because you want it to be overriden that `public` was an intentional decision here,; has `file` been considered from the persepective of it being public before or is this the first time we have had not had it `external`?
* [ ] `rug` function looks OK, reentrancy protection via mutex OK, authed correctly.
      Tying into the above point on file being public, do we want the visibility of `rug` to match that of file? The usecase for both of these is the same - either I would expect us to be compatible with internal calls for both or neither.
* [ ] `fire` is custom logic and looks OK, reentrancy protection via mutex OK
* [ ] `_swap` and `_drop` virtual functions OK.

### KilnBase.t.sol

Looking for comprehensive testing for `file`, `rug` and `fire` as well as their failure cases. No fuzzing is used so we have to assume that fuzzing won't cause any issues.

* [ ] `file` tests OK
  * [ ] `lot` 
    * [ ] There is no check that `lot!=0` (which would prevent `fire`)  however rug function negates this issue
  * [ ]  `hop`
    * [ ] There is no check that `hop!=0` (which would allow instant sale of the entire amount to sell), however this isn't an issue since it doesn't cause loss of user funds (it just negates the purpose of Kiln)
* [ ] `rug` tests OK
  * [ ] `testRugReenter` could be moved up but that's a nit
* [ ] `fire` tests OK
  * [ ] Note: no test for if `hop` is 0 but it should have the same outcome as `testFireAfterHopNotPassed` in theory.
All tests OK

### KilnMom.sol

* [ ] Solidity version is OK, same as KilnBase
* [ ] Interfaces OK
  * [ ] RugLike OK
  * [ ] AuthorityLike OK (matches [OSM mom](https://github.com/makerdao/osm-mom/blob/f9839592ea2154d24b100dcbc6097fa3954db5de/src/OsmMom.sol#L24))
* [ ] `owner` and `authority` OK - different (owned) style to `wards` based auth so I am comparing against OSM mom
* [ ] Events OK - OsmMom uses `file` based emits it seems, rather than the (presumably newer style) events here.
* [ ] `onlyOwner` modifier OK
* [ ] `auth` modifier OK
* [ ] Constructor OK
* [ ] `isAuthorized` function (I am not too familiar with this and what `authorityLike` instances are.
      Contract authed against itself (OK)
      Owner always authed (OK)
      If no authority set, return false (OK)
      Otherwise check `canCall` (I have no clue what this does)
      Looks OK
* [ ] `setOwner` looks OK (uses emit events not notes, still OK)
* [ ] `setAuthority` OK (same as above)
* [ ] `rug` external wrapper is `auth`ed correctly and uses the set `dst` and can't be used to wrap reentrancy (called `rug` has a mutex). All looks OK to me.

### KilnMom.t.sol

Looking for comprehensive testing for `setOwner`, `setAuthority` and `rug` as well as their failure cases. 

* [ ]  `setOwner` tests OK and failure cases OK 
* [ ] `setAuthority` tests OK and failure cases OK 
* [ ] `rug` tests OK and failure cases OK 
* [ ] No tests for `isAuthorized` since it is internal (OK since `dst` is immutable anyway)

### KilnUniV3.sol

* [ ] Solidity version is OK, same as KilnBase
* [ ] Imports OK, I am skipping `TwapProduct.sol` since I don't know enough about TWAP and UniV3 to accurately audit it.
* [ ] `auth` inherited from `kilnBase.sol` which is OK.
* [ ] Interfaces **POTENTIAL ISSUE**
  * [ ] SwapRouterLike => `exactInput` is defined as `external` in the contract but in [UniSwap's interfaces](https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/interfaces/ISwapRouter.sol#L37) it is `external payable`. Not sure if this will be an issue but worth matching if this is the corresponding interface.
  * [ ] SwapRouterLike => `factory` looks OK as long as it's calling the implementation behind any swap router that uses a proxy? Tests look comprehensive enough to catch issues here.
* [ ] `ExactInputParams` struct OK (matches UniSwap)
* [ ] Variables (`scope` and `yen`) look OK, `path` looks OK
* [ ] Immutables OK
* [ ] Events OK
  * [ ] Initially I noted: "Only one `File` events type exists, where `data` is bytes, however there is [an instance](https://github.com/makerdao/dss-kiln/blob/uni_v3_twap/src/KilnUniV3.sol#L91) of the `file` function being used for `data` of type `uint256`; is this intentional or should we have two event types?"
        However the File event should be inherited from KilnBase, which I didn't notice before. All OK here.
* [ ] Constructor OK
  * [ ] **NOTE:** constructor should be given the `swapRouter` address of the implementation rather than the proxy (is this correct?). If so, I would change wording to "the address of the current Uniswap V3 swap router *implementation*"
  * [ ] `scope` and `yen` default values look reasonable
* [ ] Math (WAD) declaration - I would move this above the constructor since it's used in the constructor, but no functional difference here.
* [ ] `file` function for `path` is authed, marked as external. Looks OK.
* [ ] `file` function for `yen` and `scope` is authed, marked as public (conversation about this was in KilnBase.sol review). OK.
  * [ ] `super.file()` terminates after call (OK)
  * [ ] Inherits `File` event with `uint256` type `data` from `KilnBase.sol`
  * [ ] Scope being 0 is OK and [handled by TwapProduct.sol](https://github.com/makerdao/dss-kiln/blob/uni_v3_twap/src/uniV3/TwapProduct.sol#L66).
* [ ] `_swap` function OK
      Approval OK, copying `path` and `yen` from state to memory OK, `amountMin` calculation doesn't lose accuracy/figures to rounding, struct population OK, swap execution OK.
* [ ] `_drop` function not used; OK.
  
### KilnUniV3.t.sol

I am just looking for holes in the test coverage here - I had a brief look and it all looks ok.

File: src/KilnBase.sol
Comment (The-Arbiter) [inline src/KilnBase.sol:63]:
```suggestion
        z = x <= y ? x : y;
```

File: src/KilnUniV3.sol
Comment (The-Arbiter) [inline src/KilnUniV3.sol:24]: `exactInput` is defined as `external` in the contract but in [UniSwap's interfaces](https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/interfaces/ISwapRouter.sol#L37) it is `external payable`. Not sure if this will be an issue but worth matching if this is the corresponding interface.

```suggestion
    function exactInput(ExactInputParams calldata params) external payable returns (uint256 amountOut);
```

**Review by The-Arbiter — APPROVED** [review]
Other than the two suggestions and notes in my review, LGTM 🚀

File: src/KilnBase.sol
Comment (rockyfour) [reply src/KilnBase.sol:63]: mm yeah this varies between some of the Maker contracts, but there are at least a bunch of contracts using the current code so I think we can leave as is:

* https://github.com/makerdao/dss-auto-line/blob/65c3da68d3ad63aebdd86cc669739da594f0f55c/src/DssAutoLine.sol
* https://github.com/makerdao/dss/blob/17187f7d47be2f4c71d218785e1155474bbafe8a/src/vow.sol#L91
* https://github.com/makerdao/dss/blob/fa4f6630afb0624d04a003e920b0d71a00331d98/src/end.sol#L302

File: src/KilnUniV3.sol
Comment (rockyfour) [reply src/KilnUniV3.sol:24]: The `payable` keyword is not part of the formal function signature, and in this case not having it in the interface is better for safety as there is no intention or support for transferring raw eth (so trying to do that will fail already in compilation). So I think we can leave this as is.

**Comment by rockyfour** [comment]
@The-Arbiter thank you for the detailed review and documentation!

Small note - `swapRouter` should actually be the proxy of the router, not the implementation (it is so also in the simulation tests in `KilnUniv3.t.sol`).

See the answers above for the standalone comments.

File: src/KilnUniV3.sol
Comment (The-Arbiter) [reply src/KilnUniV3.sol:24]: No problem, this is a mistake on my part.

File: src/KilnBase.sol
Comment (The-Arbiter) [reply src/KilnBase.sol:63]: Sure, no issue here then.

**Comment by The-Arbiter** [comment]
@rockyfour thanks for addressing these - none of my other points look breaking here so I'd give it the green light now. I question of the value of having 1 & 2 for the mutex vs 0 and 1 (slightly more expensive) as it really depends on how many calls to `fire` are made.

**Pull request merged**
